torch/lib/THD/master_worker/common/CommandChannel.cpp

```diff
@@ -84,19 +87,58 @@ bool MasterCommandChannel::init() {
 }
 
 void MasterCommandChannel::sendMessage(std::unique_ptr<rpc::RPCMessage> msg, int rank) {
+  // Throw error which we get from worker.
+  if (thd::master::THDState::s_error != "") {
+    throw std::runtime_error(thd::master::THDState::s_error);
+  }
+
   if ((rank <= 0) || (rank >= _sockets.size())) {
     throw std::domain_error("sendMessage received invalid rank as parameter");
   }
 
   ::thd::sendMessage(_sockets[rank], std::move(msg));
 }
 
-std::unique_ptr<rpc::RPCMessage> MasterCommandChannel::recvMessage(int rank) {
-  if ((rank <= 0) || (rank >= _sockets.size())) {
-    throw std::domain_error("recvMessage received invalid rank as parameter");
+std::tuple<rank_type, std::string> MasterCommandChannel::recvError() {
+  if (!_poll_events) {
+    // cache poll events array, it will be reused in another `receiveError` calls
+    _poll_events.reset(new struct pollfd[_sockets.size()]);
+    for (std::size_t rank = 0; rank < _sockets.size(); ++rank) {
+      _poll_events[rank] = {
+        .fd = _sockets[rank],
+        .events = POLLIN
+      };
+    }
+  }
+
+  // cleanup
+  for (std::size_t rank = 0; rank < _sockets.size(); ++rank) {
+    _poll_events[rank].revents = 0;
+  }
+
+  SYSCHECK(::poll(_poll_events.get(), _sockets.size(), -1)) // infinite timeout
+  for (std::size_t rank = 0; rank < _sockets.size(); ++rank) {
+    if (this->_poll_events[rank].revents == 0)
+      continue;
+
+    /* If `POLLIN` is not set or there is different flag set it means that
+     * something is wrong.
+     */
+    if (_poll_events[rank].revents ^ POLLIN) {
+      _poll_events[rank].fd = -1; // mark worker as ignored
+      return std::make_tuple(rank, "Connection with worker has been closed");
+    }
+
+    // receive error
+    std::uint64_t error_length;
+    recv_bytes<std::uint64_t>(_poll_events[rank].fd, &error_length, 1);
+
+    std::unique_ptr<char[]> error(new char[error_length + 1]);
+    recv_bytes<char>(_poll_events[rank].fd, error.get(), error_length);
+    return std::make_tuple(rank, std::string(error.get(), error_length));
   }
 
-  return receiveMessage(_sockets[rank]);
+  throw std::runtime_error("Failed to receive error from worker");
```

Comment:
Wait, why do you always throw here? It seems that you can't exit this function without throwing an error, or returning one received from the worker.